### 1. 데이터 준비
scikit-learn에서 제공하는 toy dataset 중, 분류 문제에 적합한 digits과 wine, breast_cancer을 import하자.

In [1]:
# import 필요한 모듈
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

# (1) 데이터 준비
from sklearn.datasets import load_digits
from sklearn.datasets import load_wine
from sklearn.datasets import load_breast_cancer

digits = load_digits()
digits_data = digits.data # Feature Data 지정하기
digits_label = digits.target # Label Data 지정하기

print('digits labels : ', digits.target_names) # Target Names 출력해 보기
print('(data 개수, features)=',digits_data.shape) # 데이터 Describe 해 보기
print('-----------------------------------------------')

wine = load_wine()
wine_data = wine.data # Feature Data 지정하기
wine_label = wine.target # Label Data 지정하기
print('wine labels : ', wine.target_names) # Target Names 출력해 보기
print('(data 개수, features)=',wine_data.shape) # 데이터 Describe 해 보기
print('-----------------------------------------------')

breast_cancer = load_breast_cancer()
breast_cancer_data = breast_cancer.data # Feature Data 지정하기
breast_cancer_label = breast_cancer.target # Label Data 지정하기
print('breast_cancer labels : ', breast_cancer.target_names) # Target Names 출력해 보기
print('(data 개수, features)=',breast_cancer_data.shape) # 데이터 Describe 해 보기

digits labels :  [0 1 2 3 4 5 6 7 8 9]
(data 개수, features)= (1797, 64)
-----------------------------------------------
wine labels :  ['class_0' 'class_1' 'class_2']
(data 개수, features)= (178, 13)
-----------------------------------------------
breast_cancer labels :  ['malignant' 'benign']
(data 개수, features)= (569, 30)


### 2. 데이터 톺아보기

In [2]:
# (2)-1 digits 데이터 톺아보기
print(digits.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 5620
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

#### digits.DESCR 설명 (Description of DESCR)
digits 에서 64는 8X8픽셀 값을 일렬로 펴진 데이터를 뜻합니다.  
펴진 data를 reshape하면 handwritten digits 이미지를 확인할 수 있고, 이 데이터가 총 1797개 입니다.  
NIST가 제공하는 사전 처리 프로그램은 미리 인쇄된 양식에서 handwritten digits의 표준화된 비트맵을 추출하는 데 사용됐습니다.  
총 43명 중에서 30명이 훈련 세트에 기여했으며 13명이 시험 세트에 기여했습니다.  
32x32 비트맵은 4x4의 분할되지 않는 블록으로 나뉘고 각 블록에서 픽셀 수가 계산됩니다.(32/4=16)  
이렇게 하면 각 요소가 0~16 범위의 정수인 8x8의 입력 행렬이 생성됩니다.  
이 과정은 차원성을 줄이고 작은 왜곡에 불변성을 부여합니다.

In [3]:
# (2)-2. wine 데이터 톺아보기
print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

#### wine.DESCR 설명 (Description of DESCR)
wine에서 13개의 feature는 와인의 특징을 수로 표현했습니다.  
이 특징의 조합으로 3개의 class로 나누어져있다.  
이 데이터는 이탈리아의 같은 지역에서 재배되는 포도주에 대한 화학 분석 결과입니다.  
와인의 세 가지 유형에서 발견되는 다른 성분들에 대해 13가지 다른 측정치가 있습니다.  
우리가 만드는 머신러닝은 13가지 attribute(속성)이 주어지면 그에 맞게 class를 맞춰야합니다.

In [4]:
# (2)-3. breast_cancer 데이터 톺아보기
print(breast_cancer.DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

#### breast_cancer.DESCR 설명 (Description of DESCR)
breasr_cancer는 30개의 속성으로 가지고 있다. 그리고 악성과 양성 2개의 class로 나뉜다. 
악성(malignant) data는 212개, 양성(benign) data는 357개이다.  
(이때 양성은 한자 '어질 량'을 써서 치료 가능한 종양을 뜻한다.)

### 3. 문제와 정답준비
pandas 라이브러리를 이용해 2차원 배열로 문제와 정답을 나열합니다.  
문제 : 머신러닝 모델에게 입력되는 데이터. feature라고 부르기도 한다. 변수 이름으로는 X를 많이 사용한다.  
정답 : 머신러닝 모델이 맞추어야 하는 데이터. label 또는 target이라고 부르기도 한다. 변수 이름으로는 y를 많이 사용한다.

In [5]:
# (3) 문제와 정답준비

import pandas as pd

# 준비한 데이터셋을 pandas가 제공하는 DataFrame 이라는 자료형으로 변환 => 문제
# digits
digits_df = pd.DataFrame(data=digits_data, columns=digits.feature_names)

# 정답인 label 컬럼 추가(가장 오른쪽에 추가됨)
digits_df["label"] = digits.target
digits_df

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0           0.0        0.0        5.0       13.0        9.0        1.0   
1           0.0        0.0        0.0       12.0       13.0        5.0   
2           0.0        0.0        0.0        4.0       15.0       12.0   
3           0.0        0.0        7.0       15.0       13.0        1.0   
4           0.0        0.0        0.0        1.0       11.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        4.0       10.0       13.0        6.0   
1793        0.0        0.0        6.0       16.0       13.0       11.0   
1794        0.0        0.0        1.0       11.0       15.0        1.0   
1795        0.0        0.0        2.0       10.0        7.0        0.0   
1796        0.0        0.0       10.0       14.0        8.0        1.0   

      pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_7  pixel_7_0  \
0           0.0        0.0        0.0        0.0  ...        0.0        0.0   
1           0.0        0.0        0.0        0.0  ...        0.0        0.0   
2           0.0        0.0        0.0        0.0  ...        0.0        0.0   
3           0.0        0.0        0.0        8.0  ...        0.0        0.0   
4           0.0        0.0        0.0        0.0  ...        0.0        0.0   
...         ...        ...        ...        ...  ...        ...        ...   
1792        0.0        0.0        0.0        1.0  ...        0.0        0.0   
1793        1.0        0.0        0.0        0.0  ...        0.0        0.0   
1794        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1795        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1796        0.0        0.0        0.0        2.0  ...        0.0        0.0   

      pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  pixel_7_6  \
0           0.0        6.0       13.0       10.0        0.0        0.0   
1           0.0        0.0       11.0       16.0       10.0        0.0   
2           0.0        0.0        3.0       11.0       16.0        9.0   
3           0.0        7.0       13.0       13.0        9.0        0.0   
4           0.0        0.0        2.0       16.0        4.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        2.0       14.0       15.0        9.0        0.0   
1793        0.0        6.0       16.0       14.0        6.0        0.0   
1794        0.0        2.0        9.0       13.0        6.0        0.0   
1795        0.0        5.0       12.0       16.0       12.0        0.0   
1796        1.0        8.0       12.0       14.0       12.0        1.0   

      pixel_7_7  label  
0           0.0      0  
1           0.0      1  
2           0.0      2  
3           0.0      3  
4           0.0      4  
...         ...    ...  
1792        0.0      9  
1793        0.0      0  
1794        0.0      8  
1795        0.0      9  
1796        0.0      8  

[1797 rows x 65 columns]

In [6]:
# wine
wine_df = pd.DataFrame(data=wine_data, columns=wine.feature_names)

# 정답인 label 컬럼 추가(가장 오른쪽에 추가됨)
wine_df["label"] = wine.target
wine_df

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  label  
0                            3.92   1065.0      0  
1                            3.40   1050.0      0  
2                            3.17   1185.0      0  
3                            3.45   1480.0      0  
4                            2.93    735.0      0  
..                            ...      ...    ...  
173                          1.74    740.0      2  
174                          1.56    750.0      2  
175                          1.56    835.0      2  
176                          1.62    840.0      2  
177                          1.60    560.0      2  

[178 rows x 14 columns]

In [7]:
# breast_cancer
breast_cancer_df = pd.DataFrame(data=breast_cancer_data, columns=breast_cancer.feature_names)

# 정답인 label 컬럼 추가(가장 오른쪽에 추가됨)
breast_cancer_df["label"] = breast_cancer.target
breast_cancer_df

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  label  
0                  0.2654          0.4601                  0.11890      0  
1                  0.1860          0.2750                  0.08902      0  
2                  0.2430          0.3613                  0.08758      0  
3                  0.2575          0.6638                  0.17300      0  
4                  0.1625          0.2364                  0.07678   

In [8]:
# 중략된 표에 1인 label이 하나 밖에 보이지 않지만 중간을 살펴보면 뒤섞여있음을 볼 수 있다.
breast_cancer_df[200:300]

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
200        12.23         19.56           78.54      461.0          0.09586   
201        17.54         19.32          115.10      951.6          0.08968   
202        23.29         26.67          158.90     1685.0          0.11410   
203        13.81         23.75           91.56      597.8          0.13230   
204        12.47         18.60           81.09      481.9          0.09965   
..           ...           ...             ...        ...              ...   
295        13.77         13.27           88.06      582.7          0.09198   
296        10.91         12.35           69.14      363.7          0.08518   
297        11.76         18.14           75.00      431.1          0.09968   
298        14.26         18.17           91.22      633.1          0.06576   
299        10.51         23.09           66.85      334.2          0.10150   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
200           0.08087         0.04187              0.04107         0.1979   
201           0.11980         0.10360              0.07488         0.1506   
202           0.20840         0.35230              0.16200         0.2200   
203           0.17680         0.15580              0.09176         0.2251   
204           0.10580         0.08005              0.03821         0.1925   
..                ...             ...                  ...            ...   
295           0.06221         0.01063              0.01917         0.1592   
296           0.04721         0.01236              0.01369         0.1449   
297           0.05914         0.02685              0.03515         0.1619   
298           0.05220         0.02475              0.01374         0.1635   
299           0.06797         0.02495              0.01875         0.1695   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
200                 0.06013  ...          28.36            92.15       638.4   
201                 0.05491  ...          25.84           139.50      1239.0   
202                 0.06229  ...          32.68           177.00      1986.0   
203                 0.07421  ...          41.85           128.50      1153.0   
204                 0.06373  ...          24.64            96.05       677.9   
..                      ...  ...            ...              ...         ...   
295                 0.05912  ...          16.93            94.17       661.1   
296                 0.06031  ...          14.82            72.42       392.2   
297                 0.06287  ...          23.39            85.10       553.6   
298                 0.05586  ...          25.26           105.80       819.7   
299                 0.06556  ...          24.22            70.10       362.7   

     worst smoothness  worst compactness  worst concavity  \
200           0.14290            0.20420          0.13770   
201           0.13810            0.34200          0.35080   
202           0.15360            0.41670          0.78920   
203           0.22260            0.52090          0.46460   
204           0.14260            0.23780          0.26710   
..                ...                ...              ...   
295           0.11700            0.10720          0.03732   
296           0.09312            0.07506          0.02884   
297           0.11370            0.07974          0.06120   
298           0.09445            0.21670          0.15650   
299           0.11430            0.08614          0.04158   

     worst concave points  worst symmetry  worst fractal dimension  label  
200               0.10800          0.2668                  0.08174      1  
201               0.19390          0.2928                  0.07867      0  
202               0.27330          0.3198                  0.08762      0  
203               0.20130          0.4432                  0.10860      0  
204               0.10150          0.3014                  0.08750   

In [9]:
# label 개수를 확인해서 각 라벨이 악성인지 음성인지 확인할 수 있다.
breast_cancer_df[breast_cancer_df.label == 0] # 212개는 악성

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
563        20.92         25.09          143.00     1347.0          0.10990   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
563           0.22360         0.31740              0.14740         0.2149   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
563                 0.06879  ...          29.41           179.10      1819.0   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   

     worst smoothness  worst compactness  worst concavity  \
0              0.1622             0.6656           0.7119   
1              0.1238             0.1866           0.2416   
2              0.1444             0.4245           0.4504   
3              0.2098             0.8663           0.6869   
4              0.1374             0.2050           0.4000   
..                ...                ...              ...   
563            0.1407             0.4186           0.6599   
564            0.1410             0.2113           0.4107   
565            0.1166             0.1922           0.3215   
566            0.1139             0.3094           0.3403   
567            0.1650             0.8681           0.9387   

     worst concave points  worst symmetry  worst fractal dimension  label  
0                  0.2654          0.4601                  0.11890      0  
1                  0.1860          0.2750                  0.08902      0  
2                  0.2430          0.3613                  0.08758      0  
3                  0.2575          0.6638                  0.17300      0  
4                  0.1625          0.2364                  0.07678   

In [10]:
breast_cancer_df[breast_cancer_df.label == 1] # 357개는 양성 (암 아님)

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
19        13.540         14.36           87.46      566.3          0.09779   
20        13.080         15.71           85.63      520.0          0.10750   
21         9.504         12.44           60.34      273.9          0.10240   
37        13.030         18.42           82.61      523.8          0.08983   
46         8.196         16.84           51.71      201.9          0.08600   
..           ...           ...             ...        ...              ...   
558       14.590         22.68           96.39      657.1          0.08473   
559       11.510         23.93           74.52      403.5          0.09261   
560       14.050         27.15           91.38      600.4          0.09929   
561       11.200         29.37           70.67      386.0          0.07449   
568        7.760         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
19            0.08129         0.06664             0.047810         0.1885   
20            0.12700         0.04568             0.031100         0.1967   
21            0.06492         0.02956             0.020760         0.1815   
37            0.03766         0.02562             0.029230         0.1467   
46            0.05943         0.01588             0.005917         0.1769   
..                ...             ...                  ...            ...   
558           0.13300         0.10290             0.037360         0.1454   
559           0.10210         0.11120             0.041050         0.1388   
560           0.11260         0.04462             0.043040         0.1537   
561           0.03558         0.00000             0.000000         0.1060   
568           0.04362         0.00000             0.000000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
19                  0.05766  ...          19.26            99.70       711.2   
20                  0.06811  ...          20.49            96.09       630.5   
21                  0.06905  ...          15.66            65.13       314.9   
37                  0.05863  ...          22.81            84.46       545.9   
46                  0.06503  ...          21.96            57.26       242.2   
..                      ...  ...            ...              ...         ...   
558                 0.06147  ...          27.27           105.90       733.5   
559                 0.06570  ...          37.16            82.28       474.2   
560                 0.06171  ...          33.17           100.20       706.7   
561                 0.05502  ...          38.30            75.19       439.6   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
19            0.14400            0.17730          0.23900   
20            0.13120            0.27760          0.18900   
21            0.13240            0.11480          0.08867   
37            0.09701            0.04619          0.04833   
46            0.12970            0.13570          0.06880   
..                ...                ...              ...   
558           0.10260            0.31710          0.36620   
559           0.12980            0.25170          0.36300   
560           0.12410            0.22640          0.13260   
561           0.09267            0.05494          0.00000   
568           0.08996            0.06444          0.00000   

     worst concave points  worst symmetry  worst fractal dimension  label  
19                0.12880          0.2977                  0.07259      1  
20                0.07283          0.3184                  0.08183      1  
21                0.06227          0.2450                  0.07773      1  
37                0.05013          0.1987                  0.06169      1  
46                0.02564          0.3105                  0.07409   

In [11]:
# target_names을 활용해서도 알 수 있다.
from sklearn.datasets import load_breast_cancer
data = load_breast_cancer()
list(data.target_names)

['malignant', 'benign']

### 4. train, test 데이터 분리
scikit-learn이 제공하는 train_test_split 이라는 함수로 데이터를 분리한 후 학습시킵니다.

In [12]:
# (4) train, test 데이터 분리
X_digits_train, X_digits_test, y_digits_train, y_digits_test = train_test_split(digits_data, digits_label, test_size=0.2, random_state=7)
print('digits_______________________________________________')
print('X_digits_train 개수:', len(X_digits_train), ', X_digits_test 개수:', len(X_digits_test))

X_wine_train, X_wine_test, y_wine_train, y_wine_test = train_test_split(wine_data, wine_label, test_size=0.2, random_state=7)
print('''
wine___________________________________________''')
print('X_wine_train 개수:', len(X_wine_train), ', X_wine_test 개수:', len(X_wine_test))

X_breast_cancer_train, X_breast_cancer_test, y_breast_cancer_train, y_breast_cancer_test = train_test_split(breast_cancer_data, breast_cancer_label, test_size=0.2, random_state=7)
print('''
breast_cancer_____________________________________________________''')
print('X_breast_cancer_train 개수:', len(X_breast_cancer_train), ', X_breast_cancer_test 개수:', len(X_breast_cancer_test))

digits_______________________________________________
X_digits_train 개수: 1437 , X_digits_test 개수: 360

wine___________________________________________
X_wine_train 개수: 142 , X_wine_test 개수: 36

breast_cancer_____________________________________________________
X_breast_cancer_train 개수: 455 , X_breast_cancer_test 개수: 114


### 5. 다양한 모델로 학습 & 평가 (Separately 1.digits/ 2. wine/ 3.breast_cancer)  

오차 행렬(confusion matrix)은 original label과 predicted label 관계를 나타낸 행렬이다.  
Classification_report는 오차 행렬을 가지고 모델을 평가할 수 있는 척도인 precision과 recall, f1-score, accuracy 총 4가지를 계산한 것이다.  

In [13]:
# 1.digits
from sklearn.metrics import classification_report
# (5)-1.Decision Tree
from sklearn.tree import DecisionTreeClassifier
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_digits_train, y_digits_train)
y_digits_pred = decision_tree.predict(X_digits_test)
print('Decision Tree_________________________________________')

# Confusion matrix(오차 행렬)
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_digits_test, y_digits_pred))

print(classification_report(y_digits_test, y_digits_pred))

# (5)-2.Random Forest 사용해 보기
from sklearn.ensemble import RandomForestClassifier
random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_digits_train, y_digits_train)
y_digits_pred = random_forest.predict(X_digits_test)
print('Random Forest_________________________________________')

# Confusion matrix(오차 행렬)
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_digits_test, y_digits_pred))

print(classification_report(y_digits_test, y_digits_pred))

# (5)-3.SVM 사용해 보기
from sklearn import svm
svm_model = svm.SVC(random_state=32)
svm_model.fit(X_digits_train, y_digits_train)
y_digits_pred = svm_model.predict(X_digits_test)
print('SVM___________________________________________________')

# Confusion matrix(오차 행렬)
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_digits_test, y_digits_pred))

print(classification_report(y_digits_test, y_digits_pred))

# (5)-4.SGD Classifier 사용해 보기
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier(random_state=32)
sgd_model.fit(X_digits_train, y_digits_train)
y_digits_pred = sgd_model.predict(X_digits_test)
print('SGD Classifier________________________________________')

# Confusion matrix(오차 행렬)
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_digits_test, y_digits_pred))

print(classification_report(y_digits_test, y_digits_pred))

# (5)-5.Logistic Regression 사용해 보기 
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()
logistic_model.fit(X_digits_train, y_digits_train)
y_digits_pred = logistic_model.predict(X_digits_test)
print('Logistic Regression___________________________________')

# Confusion matrix(오차 행렬)
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_digits_test, y_digits_pred))

print(classification_report(y_digits_test, y_digits_pred))

Decision Tree_________________________________________
[[42  0  0  1  0  0  0  0  0  0]
 [ 0 34  3  1  0  1  1  0  0  2]
 [ 0  0 33  2  0  0  1  1  2  1]
 [ 0  1  0 31  0  0  0  0  1  1]
 [ 0  0  1  0 35  0  0  0  1  0]
 [ 0  1  0  0  0 27  0  0  0  0]
 [ 0  0  0  0  2  0 26  0  0  0]
 [ 0  0  0  1  2  1  0 27  0  2]
 [ 0  5  4  1  1  0  3  0 28  1]
 [ 0  1  1  2  2  1  0  0  0 25]]
              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.81      0.81      0.81        42
           2       0.79      0.82      0.80        40
           3       0.79      0.91      0.85        34
           4       0.83      0.95      0.89        37
           5       0.90      0.96      0.93        28
           6       0.84      0.93      0.88        28
           7       0.96      0.82      0.89        33
           8       0.88      0.65      0.75        43
           9       0.78      0.78      0.78        32

    accuracy      

/home/aiffel/Downloads/yes/envs/aiffel/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


### 6-1. digits 평가 
digits 결과 평가의 기준을 precision(정밀도), recall(재현율), f1-score(재현율과 정밀도의 조화평균), accuracy(정확도) 중 어느 것으로 하던지 간에 SVM이 0.99로 가장 높다.  
SVM은 데이터를 고차원으로 보내서 선형 분리하는 모델이고 매우 엄격하게 데이터를 분리한다.  
비교적 간단한 2차원 이미지 데이터를 고차원으로 보내서 분리한 것이 높은 정확도를 냈다고 예상된다.  

In [14]:
# 2.wine
from sklearn.metrics import classification_report
# (5)-1.Decision Tree
from sklearn.tree import DecisionTreeClassifier
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_wine_train, y_wine_train)
y_wine_pred = decision_tree.predict(X_wine_test)
print('Decision Tree_________________________________________')

# Confusion matrix(오차 행렬)
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_wine_test, y_wine_pred))

print(classification_report(y_wine_test, y_wine_pred))

# (5)-2.Random Forest 사용해 보기
from sklearn.ensemble import RandomForestClassifier
random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_wine_train, y_wine_train)
y_wine_pred = random_forest.predict(X_wine_test)
print('Random Forest_________________________________________')

# Confusion matrix(오차 행렬)
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_wine_test, y_wine_pred))

print(classification_report(y_wine_test, y_wine_pred))

# (5)-3.SVM 사용해 보기
from sklearn import svm
svm_model = svm.SVC(random_state=32)
svm_model.fit(X_wine_train, y_wine_train)
y_wine_pred = svm_model.predict(X_wine_test)
print('SVM___________________________________________________')

# Confusion matrix(오차 행렬)
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_wine_test, y_wine_pred))


print(classification_report(y_wine_test, y_wine_pred))

# (5)-4.SGD Classifier 사용해 보기
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier(random_state=32)
sgd_model.fit(X_wine_train, y_wine_train)
y_wine_pred = sgd_model.predict(X_wine_test)
print('SGD Classifier________________________________________')

# Confusion matrix(오차 행렬)
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_wine_test, y_wine_pred))


print(classification_report(y_wine_test, y_wine_pred))

# (5)-5.Logistic Regression 사용해 보기 
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()
logistic_model.fit(X_wine_train, y_wine_train)
y_wine_pred = logistic_model.predict(X_wine_test)
print('Logistic Regression___________________________________')

# Confusion matrix(오차 행렬)
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_wine_test, y_wine_pred))


print(classification_report(y_wine_test, y_wine_pred))

Decision Tree_________________________________________
[[ 7  0  0]
 [ 0 17  0]
 [ 0  2 10]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       0.89      1.00      0.94        17
           2       1.00      0.83      0.91        12

    accuracy                           0.94        36
   macro avg       0.96      0.94      0.95        36
weighted avg       0.95      0.94      0.94        36

Random Forest_________________________________________
[[ 7  0  0]
 [ 0 17  0]
 [ 0  0 12]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       1.00      1.00      1.00        17
           2       1.00      1.00      1.00        12

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36

SVM___________________________________________________

/home/aiffel/Downloads/yes/envs/aiffel/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/aiffel/Downloads/yes/envs/aiffel/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


### 6-2. wine 평가 
wine 결과 평가의 기준도 digits과 비슷하게 precision(정밀도), recall(재현율), f1-score(재현율과 정밀도의 조화평균), accuracy(정확도) 중 어느 것으로 하던지 간에 RandomForest model 하나가 1.00로 가장 높다.  
이름에서 유추할 수 있듯이 Decision tree의 집합체를 RandomForest라고 한다. 같은 데이터에 대해 의사결정나무를 여러 개 만들어 그 결과를 종합해 예측 성능을 높이는 기법이다.
Decision tree 에서는 class 3 를 class 2라고 잘못 분류했지만 RandomForest에서는 완벽히 분리했다.  

SVM에서는 아주 terrible한 결과가 나왔다. digits과 완전히 반대 결과이다.  
게다가 SVM은 예측이 어떻게 결정되었는지 이해하기 어렵다. 그러나 클래스 사이의 경계에 위치한 데이터 포인트들이 충분한 거리로 떨어지지 않아서라고 추측해 볼 수 있다.

In [15]:
# 3.Breast_cancer
from sklearn.metrics import classification_report
# (5)-1.Decision Tree
from sklearn.tree import DecisionTreeClassifier
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_breast_cancer_train, y_breast_cancer_train)
y_breast_cancer_pred = decision_tree.predict(X_breast_cancer_test)
print('Decision Tree_________________________________________')

# Confusion matrix(오차 행렬)
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_breast_cancer_test, y_breast_cancer_pred))

print(classification_report(y_breast_cancer_test, y_breast_cancer_pred))

# (5)-2.Random Forest 사용해 보기
from sklearn.ensemble import RandomForestClassifier
random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_breast_cancer_train, y_breast_cancer_train)
y_breast_cancer_pred = random_forest.predict(X_breast_cancer_test)
print('Random Forest_________________________________________')

# Confusion matrix(오차 행렬)
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_breast_cancer_test, y_breast_cancer_pred))


print(classification_report(y_breast_cancer_test, y_breast_cancer_pred))

# (5)-3.SVM 사용해 보기
from sklearn import svm
svm_model = svm.SVC(random_state=32)
svm_model.fit(X_breast_cancer_train, y_breast_cancer_train)
y_breast_cancer_pred = svm_model.predict(X_breast_cancer_test)
print('SVM___________________________________________________')

# Confusion matrix(오차 행렬)
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_breast_cancer_test, y_breast_cancer_pred))


print(classification_report(y_breast_cancer_test, y_breast_cancer_pred))

# (5)-4.SGD Classifier 사용해 보기
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier(random_state=32)
sgd_model.fit(X_breast_cancer_train, y_breast_cancer_train)
y_breast_cancer_pred = sgd_model.predict(X_breast_cancer_test)
print('SGD Classifier________________________________________')

# Confusion matrix(오차 행렬)
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_breast_cancer_test, y_breast_cancer_pred))


print(classification_report(y_breast_cancer_test, y_breast_cancer_pred))

# (5)-5.Logistic Regression 사용해 보기 
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()
logistic_model.fit(X_breast_cancer_train, y_breast_cancer_train)
y_breast_cancer_pred= logistic_model.predict(X_breast_cancer_test)
print('Logistic Regression___________________________________')

# Confusion matrix(오차 행렬)
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_breast_cancer_test, y_breast_cancer_pred))


print(classification_report(y_breast_cancer_test, y_breast_cancer_pred))

Decision Tree_________________________________________
[[33  7]
 [ 3 71]]
              precision    recall  f1-score   support

           0       0.92      0.82      0.87        40
           1       0.91      0.96      0.93        74

    accuracy                           0.91       114
   macro avg       0.91      0.89      0.90       114
weighted avg       0.91      0.91      0.91       114

Random Forest_________________________________________
[[40  0]
 [ 0 74]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        40
           1       1.00      1.00      1.00        74

    accuracy                           1.00       114
   macro avg       1.00      1.00      1.00       114
weighted avg       1.00      1.00      1.00       114

SVM___________________________________________________
[[29 11]
 [ 0 74]]
              precision    recall  f1-score   support

           0       1.00      0.72      0.84        40
           1   

/home/aiffel/Downloads/yes/envs/aiffel/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


### 6-3. breast_cancer평가 
breast_cancer 결과는 특히 FP보다 TN이 훨씬 적게 나와야한다. (양성 종양을 잘못 분류하는 일보다 악성 종양을 잘못 분류해 양성으로 나오는 일이 적어야 한다.)  
즉, 모델 결과는 recall로 비교할 것이다. recall(재현율)은 RandomForest model에서 1.00로 가장 높다.  
이름에서 유추할 수 있듯이 Decision tree의 집합체를 RandomForest라고 한다. 같은 데이터에 대해 의사결정나무를 여러 개 만들어 그 결과를 종합해 예측 성능을 높이는 기법이다.
Decision tree 에서는 10개, 약 9%의 data를 잘못 분류했지만 RandomForest에서는 완벽히 분리했다.  

SGD Classifier가 상대적으로 recall평가 점수가 가장 낮다. 데이터를 하나씩 처리하면서 하강하기 때문에 그 사이에 생긴 miss-match라고 추측해볼 수  있다.

### 7. 고찰 Consideration
초반에는 5개나 되는 분류 모델을 배우려니 힘들기도 했다. 그러나 어느 정도 익숙해지고 나니 비교하며 볼 수 있어서 더 많은 이해를 할 수 있었다.
어렴풋이 알고 있던 오차행렬을 복습하고 breast_cancer 평가에서 활용할 수 있어서 더 확실히 배울 수 있었다.  
  
그리고 breast_cancer에서 label 0과 1 중 어떤 것이 악성이고 양성인지를 확인할 지 고민하다가 라벨이 0인 것과 1인 것만 따로 출력해서 숫자로 추측했다.   
하면서도 어딘가 모르게 찝찝했는데 팀원이 1.데이터 준비하기에 있던 target_names 상기시켜줘서 시원하게 마무리 지을 수 있었다.  

그리고 Logistic Regression에서 3번 모두 error가 떴다. 그 이유는 data 개수(data scale)이 적어서 나는 error이다. 이 error는 warning의 의미여서 결과에 많은 영향을 주는 건 아닌 것 같다.